### ISSUE TITLE
List Comprehension

### ISSUE THREAD

**Issue by tohrnii** [0xPolygonMiden/air-script#87]
It would be nice to have python style [list comprehension](https://www.geeksforgeeks.org/python-list-comprehension/) supported in AirScript. For eg.

```
trace_columns:
  main: [a, b, c[4]]

# raise value in the current row to power 7
let x = [col^7 for col in c]

# raise value in the next row to power 7
let y = [col'^7 for col in c]
```

In both cases, the result would be a vector of 4 elements.

We could also add support for iterating over two lists simultaneously like:
```
trace_columns:
  main: [a, b, c[4], d[4]]

let diff = [x - y for (x, y) in (c, d)]
```

We could also add support for iterating and enumerating like:

```
trace_columns:
  main: [a, b, c[4]]

let x = [2^i * c for (i, c) in (0..3, c)]
```

We could also support list folding. There are two possible syntax:

```
trace_columns:
  main: [a, b, c[4], d[4]]

# compute sum of products of values in c and d
let x += c * d for (c, d) in (c, d)

# compute product of sums of values in c and d
let y *= c + d for (c, d) in (c, d)
```
OR

```
trace_columns:
  main: [a, b, c[4], d[4]]

# compute sum of products of values in c and d
let x = sum([c * d for (c, d) in (c, d)])

# compute product of sums of values in c and d
let y = prod([c + d for (c, d) in (c, d)])
```
_The syntax was proposed by @bobbinth [here](https://github.com/maticnetwork/miden/discussions/254#discussioncomment-3293836)._

**Comment by tohrnii** [comment]
Question:
Should we allow users to access the length of the array for iterators? So, for eg something like,
```
trace_columns:
  main: [a, b, c[4]]

let x = [2^i * c for (i, c) in (c.len(), c)]
```

instead of 
```
trace_columns:
  main: [a, b, c[4]]

let x = [2^i * c for (i, c) in (0..3, c)]
```

**Comment by tohrnii** [comment]
Another question: Do we have consensus on using `0..n` for ranges instead of pythonic `range(0, n)`?

**Comment by bobbinth** [comment]
> Should we allow users to access the length of the array for iterators/ enumeration or in general?

I can't think of cases where we'd need something like this - but I might not be thinking of everything here. In general, my preference would not to add it until we have a clear use-case. So, unless it is super-trivial to add, I'd probably skip it for now.

> Do we have consensus on using `0..n` for ranges instead of pythonic `range(0, n)`?

I have a slight preference for `0..n` as I think it improves readability.

**Comment by tohrnii** [comment]
Another question: Which of the two syntax should we use for list folding? I think `sum()` and `prod()` seems clearer.

**Comment by bobbinth** [comment]
> Another question: Which of the two syntax should we use for list folding? I think `sum()` and `prod()` seems clearer.

Agreed!

**Comment by tohrnii** [comment]
Closed by #119, #127, #136 and and #173.
